### PR TITLE
feat: remaining record content types to support

### PR DESCRIPTION
**Pull request by 0xmuralik**
Closes: #51 

## Description
Record attributes indexing for remaining defined type of records.

______

For contributor use:

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added  relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer

______

For admin use:

- [ ] Added appropriate labels to PR (ex. `WIP`, `R4R`, `docs`, etc)
- [ ] Reviewers assigned
- [ ] Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

### CODE REVIEW

File: proto/vulcanize/registry/v1beta1/attributes.proto
Comment (i-norden) [inline proto/vulcanize/registry/v1beta1/attributes.proto:45]: This was underspecified on my part, instead of a `raw_multicodec` field and a `raw_binary` field, all we need is a `hash_reference` field. And this reference will be a CID or multihash which uses the [raw mutliformat prefix](https://github.com/multiformats/multicodec/blob/master/table.csv#L40) and references the binary as  an IPLD object.

File: proto/vulcanize/registry/v1beta1/attributes.proto
Comment (i-norden) [inline proto/vulcanize/registry/v1beta1/attributes.proto:93]: Let's rename this to `JSPackage` and add a `string name=5;` field

File: proto/vulcanize/registry/v1beta1/attributes.proto
Comment (i-norden) [inline proto/vulcanize/registry/v1beta1/attributes.proto:94]: This will reference a registered `GitRepository` record by CID/multihash

`GitRepository` also has a `repo_reference` field which will contain a qualitative/non-content-hash-reference e.g. a github or gitea web url.

It might be helpful to signal in the naming of the various `reference` fields which references are content-hash => content references vs one's that are not. E.g. `hash_reference` vs `reference`.

File: proto/vulcanize/registry/v1beta1/attributes.proto
Comment (i-norden) [inline proto/vulcanize/registry/v1beta1/attributes.proto:100]: Let's add fields:

`string chain_id`
`string network_id`
`string genesis_hash`

File: proto/vulcanize/registry/v1beta1/attributes.proto
Comment (i-norden) [inline proto/vulcanize/registry/v1beta1/attributes.proto:57]: Note: `hash_reference` to a `GitRepository` IPLD object

File: proto/vulcanize/registry/v1beta1/attributes.proto
Comment (i-norden) [inline proto/vulcanize/registry/v1beta1/attributes.proto:56]: Note: `hash_reference` to binary IPLD object

File: proto/vulcanize/registry/v1beta1/attributes.proto
Comment (i-norden) [inline proto/vulcanize/registry/v1beta1/attributes.proto:64]: Note: `hash_reference` to a `GitRepository` IPLD object

File: proto/vulcanize/registry/v1beta1/attributes.proto
Comment (i-norden) [inline proto/vulcanize/registry/v1beta1/attributes.proto:71]: Note: `hash_reference` to a `ChainRegistrationRecord` IPLD object

File: proto/vulcanize/registry/v1beta1/attributes.proto
Comment (i-norden) [inline proto/vulcanize/registry/v1beta1/attributes.proto:75]: Note: `hash_reference` to a WASM IPLD object

File: proto/vulcanize/registry/v1beta1/attributes.proto
Comment (i-norden) [inline proto/vulcanize/registry/v1beta1/attributes.proto:86]: Note: `hash_reference` to `ServiceProviderRecord` IPLD object

File: proto/vulcanize/registry/v1beta1/attributes.proto
Comment (i-norden) [inline proto/vulcanize/registry/v1beta1/attributes.proto:88]: Note: `hash_reference` to `WatcherRegistrationRecord` IPLD object

File: proto/vulcanize/registry/v1beta1/attributes.proto
Comment (i-norden) [inline proto/vulcanize/registry/v1beta1/attributes.proto:87]: Note: `hash_reference` to auction result IPLD object

File: proto/vulcanize/registry/v1beta1/attributes.proto
Comment (i-norden) [inline proto/vulcanize/registry/v1beta1/attributes.proto:95]: Note: `hash_reference` to `JSPackage` IPLD object

**Review by i-norden — CHANGES_REQUESTED** [review]
This all looks great! I made some comments to fill in a few fields that were missing in the specification or under/mis-specified, also suggest clarifying which fields are content hash references (renaming to e.g. `hash_reference`) so went through and noted all of the linkages.

**Comment by i-norden** [comment]
@0xmuralik for all of the `reference` fields that are hash-links, we should considering using the previous convention with "/"

E.g. `npm_package_ref` renamed to `hash_reference` (or not, since now the "/" already signifies the hash-link nature of the reference) with `hash_reference` a map of "/" to the CID. This should prevent the need for https://github.com/cerc-io/laconicd/issues/78.

File: proto/vulcanize/registry/v1beta1/attributes.proto
Comment (0xmuralik) [reply proto/vulcanize/registry/v1beta1/attributes.proto:94]: There is also a npm_package_reference (CID) in this message. Should we go with renaming repo_reference with repo_hash_reference and npm_package_reference with js_package_hash_ref?

File: proto/vulcanize/registry/v1beta1/attributes.proto
Comment (0xmuralik) [reply proto/vulcanize/registry/v1beta1/attributes.proto:100]: is the genesis_hash a CID reference or just a hash string?

File: proto/vulcanize/registry/v1beta1/attributes.proto
Comment (0xmuralik) [reply proto/vulcanize/registry/v1beta1/attributes.proto:88]: Is the auction_ref too an IPLD hash_reference?

File: proto/vulcanize/registry/v1beta1/attributes.proto
Comment (0xmuralik) [inline proto/vulcanize/registry/v1beta1/attributes.proto:133]: @i-norden Can you suggest a preffered method here?

File: proto/vulcanize/registry/v1beta1/attributes.proto
Comment (0xmuralik) [inline proto/vulcanize/registry/v1beta1/attributes.proto:176]: @i-norden Can you suggest a preffered method here?

File: proto/vulcanize/registry/v1beta1/attributes.proto
Comment (i-norden) [reply proto/vulcanize/registry/v1beta1/attributes.proto:104]: Good question... let's go with full CID for now

File: proto/vulcanize/registry/v1beta1/attributes.proto
Comment (i-norden) [reply proto/vulcanize/registry/v1beta1/attributes.proto:133]: Same here unfortunately, not familiar with these types and isn't obvious to me what `ref` and `os` are supposed to represent in these mappings. Going to ask Ashwin.

File: proto/vulcanize/registry/v1beta1/attributes.proto
Comment (i-norden) [reply proto/vulcanize/registry/v1beta1/attributes.proto]: We can avoid adding `hash` and instead we should revert to using a `/` sub-mapping to denote CID/link

**Review by i-norden — COMMENTED** [review]
Thanks Murali! Before we merge this, we need to make the corresponding changes in laconic-sdk and laconic-cli-registry and run their tests to ensure these changes work with those.

Let's hold off in this for now as Michael is working on squashing some intercompatibility bugs due to not doing this in the past, and it will be cleaner to work off of his results than to begin in parallel.

**Comment by 0xmuralik** [comment]
TODO: Merge https://github.com/cerc-io/laconic-sdk/pull/27

**Review by i-norden — CHANGES_REQUESTED** [review]
@0xmuralik I haven't dug into this again fully yet, but before doing so let's revert back to the
"/" usage for link delineation in order to fix this bug: https://github.com/cerc-io/laconicd/issues/78 (and also will be useful elsewhere).

**Comment by 0xmuralik** [comment]
> @0xmuralik I haven't dug into this again fully yet, but before doing so let's revert back to the "/" usage for link delineation in order to fix this bug: #78 (and also will be useful elsewhere).

To use "/" we'll have to use a map for hash references. As using maps in cosmos-sdk might lead to non determinism and is not recommended, I have taken this approach. Even though there will be only one item in this map of references and we don't have any risk of non-determinisim, I thought its safer to avoid maps and just change the getReference logic in GQL to support the new `HashReference` type.

**Comment by i-norden** [comment]
@0xmuralik let's revert to using "/" please, it was a mistake on my part to suggest we move away from this standard in the first place. I appreciate the concerns for non-determinism, but as you said since it only contains a single entry this is not an issue. This will fix the issue in https://github.com/cerc-io/laconicd/issues/78 and will prevent similar issues from cropping up downstream. In general, we need some way of detecting whether or not a field in these records is a link without knowing the the specifics of the field (e.g. name), this will be useful once we move to more generic type support. Using "/" to delineate a link is also the standard in the IPLD domain.

**Comment by 0xmuralik** [comment]
@i-norden @aleem1314 getting error while using map[] in txs
`vulcanize.registry.v1beta1.WebsiteRegistrationRecord.RepoReferenceEntry"
  does not implement proto.Message` at
https://github.com/cosmos/cosmos-sdk/blob/f71df80e93bffbf7ce5fbd519c6154a2ee9f991b/x/auth/tx/decoder.go#L40

Same error is returned when @aleem1314 tried with a unit test inculing map[] for https://github.com/cosmos/cosmos-sdk/blob/f71df80e93bffbf7ce5fbd519c6154a2ee9f991b/codec/unknownproto/unknown_fields.go#L41

**Comment by 0xmuralik** [comment]
Submitted issue to cosmos-sdk
https://github.com/cosmos/cosmos-sdk/issues/15254

**Comment by i-norden** [comment]
Thanks for opening that issue, looks like I may have sent you on a wild goose hunt. Even if the change is acceptable I don't think we can expect the ICF team to investigate and implement the change as it is not critical to the sdk. So I think we have two options: 

1. Figure out the fix and make a patch to cosmos-sdk and attempt to upstream ourselves. I think this is the only way we can reasonably expect a change like this to make it upstream.
2. Figure out some other way of demarcating Links in these object. Need to give this some more thought.

**Comment by 0xmuralik** [comment]
@i-norden by adding a json tag to the ref field , we can avoid using maps and still keep "/" for referencing CIDs.
The proto messages will use the `Hash reference` message. When the message is converted to and from json it will use a "/" instead of "ref" as the key to CID reference.
The store will have record objects with "Hash Reference" proto message but the attribute keys used for gql queries will use "/" for CIDs.

**Comment by 0xmuralik** [comment]
See https://github.com/cerc-io/laconic-sdk/pull/36.

**Pull request closed without merging**
